redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+
+    def update(self, x, who=None):
+        L = self.buffers[self.children.index(who)]
+        L.append(x)
+        if len(L) == 1 and all(self.buffers):
+            if self.prior:
+                for i in range(len(self.buffers)):
+                    # If the docs don't match, preempt with prior good result
+                    if self.buffers[i][0][0] != self.buffers[0][0][0]:
+                        self.buffers[i].appendleft(self.prior[i])
+            tup = tuple(buf.popleft() for buf in self.buffers)
+            self.condition.notify_all()
+            self.prior = tup
+            return self.emit(tup)
+        elif len(L) > self.maxsize:
+            return self.condition.wait()
+
+
+class bundle(EventStream):
+    """Combine multiple event streams into one"""
+    def __init__(self, *children, **kwargs):
+        """Initialize the Node
+
+        Parameters
+        ----------
+        children: EventStream instances
+            The event streams to be zipped together
+        """
+        self.maxsize = kwargs.pop('maxsize', 100)
+        self.buffers = [deque() for _ in children]
+        self.condition = Condition()
+        self.prior = ()
+        EventStream.__init__(self, children=children)
+        self.generate_provenance()
+
+    def update(self, x, who=None):
+        L = self.buffers[self.children.index(who)]
+        L.append(x)
+        if len(L) == 1 and all(self.buffers):
+            # if all the docs are of the same type and not an event, issue
+            # new documents which are combined
+            rvs = []
+            while all(self.buffers):
+                if all([b[0][0] == self.buffers[0][0][0] and b[0][0] != 'event'
+                        for b in self.buffers]):
+                    res = self.dispatch(
+                        tuple([b.popleft() for b in self.buffers]))
+                    rvs.append(self.emit(res))
+                elif any([b[0][0] == 'event' for b in self.buffers]):
+                    for b in self.buffers:
+                        while b:
+                            nd_pair = b[0]
+                            # run the buffers down until no events are left
+                            if nd_pair[0] != 'event':
+                                break
+                            else:
+                                nd_pair = b.popleft()
+                                new_nd_pair = super().event(
+                                    self.refresh_event(nd_pair[1]))
+                                rvs.append(self.emit(new_nd_pair))
+
+                else:
+                    raise RuntimeError("There is a mismatch of docs, but none "
+                                       "of them are events so we have reached "
+                                       "a potential deadlock, so we raise "
+                                       "this error instead")
+
+            return rvs
+        elif len(L) > self.maxsize:
+            return self.condition.wait()
```

Comment:
maybe `if all(self.buffers) return ValueError` or something. Shouldn't happen but it would be good to flag it just in case.